### PR TITLE
feat(perf-issue): make use of multiple groups in post_process_group

### PR DESCRIPTION
**Pull request by barkbarkimashark**
We've introduced the plumbing in post_process_group to take in multiple groups associated with an event. 

This PR begins to make use of these multiple groups by modifying `post_process_group` to apply logic to multiple groups before deleting the event from the cache.

### CODE REVIEW

**Review by wedamija — COMMENTED** [review]
Just a first pass review, looking pretty good. Will go through in more detail after lunch

File: src/sentry/tasks/post_process.py [hunk 1]
```diff
@@ -405,20 +405,34 @@ def post_process_group(
...
-        bind_organization_context(event.project.organization)
+        group_events: Mapping[int, GroupEvent] = {
+            ge.group_id: ge for ge in list(event.build_group_events())
+        }
+
+        multi_groups: Mapping[int, Tuple[GroupEvent, GroupState]] = {
+            gs.get("id"): (group_events.get(gs.get("id")), gs)
+            for gs in group_states
+            if gs.get("id") is not None
+        }
```
Comment (rjo100) [inline src/sentry/tasks/post_process.py:421]: does this need to be in a dict if we're going to ignore the key for `group_jobs`?

File: src/sentry/tasks/post_process.py [hunk 2]
```diff
@@ -403,20 +403,34 @@ def post_process_group(
...
+                "has_reappeared": bool(not v[1]["is_new"]),
             }
+            for k, v in multi_groups.items()
+        ]
+
+        for job in group_jobs:
             run_post_process_job(job)
 
+        with metrics.timer("tasks.post_process.delete_event_cache"):
+            event_processing_store.delete_by_key(cache_key)
```
[the remark this answers is not in the document]
Comment (rjo100) [reply src/sentry/tasks/post_process.py]: wrap it in a `try...finally`? seems kludgy thoug

File: src/sentry/tasks/post_process.py [hunk 3]
```diff
@@ -456,25 +468,38 @@ def process_event(data: dict, group_id: Optional[int]) -> Event:
...
-    # stats.
-    with sentry_sdk.start_span(op="tasks.post_process_group.fetch_buffered_group_stats"):
-        fetch_buffered_group_stats(event.group)
-
-    event.group.project = event.project
-    event.group.project.set_cached_field_value("organization", event.project.organization)
+    group_states = group_states or ([{"id": event.group_id}] if event.group_id else [])
+    rebound_groups = []
+    for group_state in group_states:
+        if group_state.get("id"):
```
Comment (wedamija) [inline src/sentry/tasks/post_process.py:488]: Are there cases where a `GroupState` won't have an id?

File: src/sentry/tasks/post_process.py [hunk 4]
```diff
@@ -456,25 +468,38 @@ def process_event(data: dict, group_id: Optional[int]) -> Event:
...
+    event.groups = rebound_groups
+    for group in event.groups:
+        # We fetch buffered updates to group aggregates here and populate them on the Group. This
+        # helps us avoid problems with processing group ignores and alert rules that rely on these
+        # stats.
+        with sentry_sdk.start_span(op="tasks.post_process_group.fetch_buffered_group_stats"):
+            fetch_buffered_group_stats(group)
+
+        group.project = event.project
+        group.project.set_cached_field_value("organization", event.project.organization)
```
Comment (wedamija) [inline src/sentry/tasks/post_process.py:502]: Wondering if it makes sense to include this logic in the first loop

File: src/sentry/tasks/post_process.py [hunk 5]
```diff
@@ -387,36 +391,44 @@ def post_process_group(
...
-
         # TODO: Remove this check once we're sending all group ids as `group_states` and treat all
         # events the same way
-        if event.get_event_type() != "transaction" or group_states is None:
+        if not is_transaction_event or group_states is None:
             # error issue
             group_states = [
                 {
+                    # have to use the event.group_id field, instead of the passed in group_id
+                    # since we rebind it in `update_event_groups` (for merging groups)
```
Comment (wedamija) [inline src/sentry/tasks/post_process.py:401]: I think we changed this back, let's remove the comment

**Review by wedamija — APPROVED** [review]
Looks good, comments are just minor things

**Pull request merged**